### PR TITLE
Add execute until functionality

### PR DESCRIPTION
**Pull request by fredthomsen**
Add Job.until function which will execute a job until a certain time is
passed.

### CODE REVIEW

**Comment by newbazz** [comment]
Please write the issue reference.

**Comment by fredthomsen** [comment]
@newbazz partially fulfills #132.

**Comment by fredthomsen** [comment]
@newbazz or @dbader anything I can do to get this merged in, or is this not desirable?

**Comment by fredthomsen** [comment]
Bumping this after a few years. @dbader anything I can do here to get this merged in?

File: schedule/__init__.py [hunk 1]
```diff
@@ -465,6 +486,9 @@ def _schedule_next_run(self):
             if (self.next_run - datetime.datetime.now()).days >= 7:
                 self.next_run -= self.period
 
+        if self.final_run and self.next_run > self.final_run:
```
Comment (SijmenHuizenga) [inline schedule/__init__.py:489]: I would rather see this check implemented in the `run()` method.

File: schedule/__init__.py [hunk 2]
```diff
@@ -465,6 +486,9 @@ def _schedule_next_run(self):
...
+            self.next_run = datetime.datetime.max
```
Comment (SijmenHuizenga) [inline schedule/__init__.py:490]: If we write this check in the `run()` method, we can return `CancelJob`, which in turn will make the scheduler unschedule the job all together. That way we don't need `datetime.datetime.max`.

File: test_schedule.py [hunk 3]
```diff
@@ -120,6 +120,10 @@ def test_next_run_time(self):
             assert every().friday.do(mock_job).next_run.day == 8
             assert every().saturday.do(mock_job).next_run.day == 9
             assert every().sunday.do(mock_job).next_run.day == 10
+            assert every().minute.until('12:17').do(
+                mock_job).next_run.minute == 16
+            assert every().minute.until('2010-01-06 09:00').do(
+                mock_job).next_run.year == 9999
```
Comment (SijmenHuizenga) [inline test_schedule.py:126]: I think we can add some more test cases here. For example: what would happen if someone enters an invalid date, for example `33:77`?

And we need to add tests to validate that a job is actually canceled after the deadline is reached.

File: schedule/__init__.py [hunk 4]
```diff
@@ -368,6 +369,26 @@ def to(self, latest):
         self.latest = latest
         return self
 
+    def until(self, datetime_str):
+        """
+        Schedule job will run until the specified time string.
+
+        :param datetime_str: A string in `HH:MM` format for same day or
+            `YYYY-MM-DD HH:MM` for a different day representing the latest
```
Comment (SijmenHuizenga) [inline schedule/__init__.py:377]: Should we support second-precision?

**Review by SijmenHuizenga — CHANGES_REQUESTED** [review]
Thanks for creating this pr! This seems like a good feature to have, and you are on the right track to implement it. Before we get into the code itself, let's take a closer look at the expected behaviour. Let's take this example code:
```
schedule.every(10).seconds.until("23:30").do(job)
while True:
    scheduler.run_pending()
    time.sleep(24)
```
Let's say this code is executed at 23:40. What will happen? I guess in the current implementation the job runs once and is canceled afterwards. Is this expected behaviour?

Another scenario: Let's say above code block starts at 23:29:00. The first execution is at 23:29:24 (scheduled execution was 23:29:10), the second execution is at 23:29:48 (scheduled execution was 23:29:34) and the third execution happens at 23:30:12 (scheduled execution was 23:29:58). At all three the times the job fires. Is this expected behaviour?

Answers to these questions could very well be answered as part of the documentation. 

Let me know if you need help with anything.

**Comment by fredthomsen** [comment]
Thanks for the review. Been a few years since I did this, I'll get it rebased and address your comments

-------- Original Message --------

File: schedule/__init__.py [hunk 5]
```diff
@@ -368,6 +369,26 @@ def to(self, latest):
...
+            time a job can be run
+        :return: The invoked job instance
+        """
+        if ' ' in datetime_str:
```
[the remark this answers is not in the document]
Comment (fredthomsen) [reply schedule/__init__.py:525]: Doing regex against the timestamp seems more painful and less accurate than just feeding various valid format strings into `datetime.datetime.strptime` and using the first valid one.

File: schedule/__init__.py (hunk 4 above)
Comment (fredthomsen) [reply schedule/__init__.py]: We should.  I'll make this update.

File: schedule/__init__.py [hunk 6]
```diff
@@ -368,6 +369,26 @@ def to(self, latest):
         self.latest = latest
         return self
 
+    def until(self, datetime_str):
+        """
+        Schedule job will run until the specified time string.
+
+        :param datetime_str: A string in `HH:MM` format for same day or
```
[the remark this answers is not in the document]
Comment (fredthomsen) [reply schedule/__init__.py]: I have an example, but shouldn't this method's docstring, which I will clean up get auto generated in the docs?

**Comment by fredthomsen** [comment]
@SijmenHuizenga I've made the updates as requested.

File: schedule/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (SijmenHuizenga) [reply schedule/__init__.py]: Ah yes! Great idea :+1:

File: schedule/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (SijmenHuizenga) [reply schedule/__init__.py]: Yes, the docstring will end up in the reference doc. The docstring is meant for detailed description of the working of the method. The examples page is there to show usage in a simpler format using examples. I will update it

**Comment by SijmenHuizenga** [comment]
You did a great job at this, thanks a lot @fredthomsen! I've just pushed some final touches. Just a summary:

* added support for accepting `datetime.timedelta` and `datetime.time`
* Moved the documentation around a bit. Examples in the rst docs and detailed explanation in the docstring.
* Moved the `convert_str_to_datetime` into its own function `_decode_datetimestr`. We can use that thing in the future to handle decoding for the `.at()` method, but that's a subject for another pr.
* Changed the `run` method to cancel a job when it is about to run and the current moment is after the deadline AND cancel it if the next scheduled timestamp is after the deadline. This way the common rule becomes "The job will never run after the specified deadline". Do you think this logic makes sense?
* Refactored the 'should this be canceled?' into the `_is_overdue` method so that we are able to extend on this when we add 'only run n times'
* Grouped the tests into functional groups.

Let me know if you think this all makes sense!

**Comment by SijmenHuizenga** [comment]
Well.. Looking at those tests I've got some work to do :sweat_smile:

**Review by SijmenHuizenga — APPROVED** [review]

**Comment by fredthomsen** [comment]
Yes, I like these changes esp. the checks in `run()`.  Makes more sense that the job can never run after `cancel_after` no matter what.  Just made an additional update to the `until` parameter type hinting.  Looks good to me.

**Pull request merged**